### PR TITLE
Support SizeOf

### PR DESCRIPTION
**Pull request by hbxie**
Partially addresses #36.

### CODE REVIEW

File: src/main/java/com/yahoo/bullet/parsing/Clause.java [hunk 1]
```diff
@@ -44,10 +44,12 @@
         @SerializedName("OR")
         OR,
         @SerializedName("NOT")
-        NOT;
+        NOT,
+        @SerializedName("SIZEOF")
+        SIZE_OF;
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/parsing/Clause.java]: Move to after REGEX_LIKE

File: src/main/java/com/yahoo/bullet/querying/FilterOperations.java [hunk 2]
```diff
@@ -95,16 +98,31 @@
         return (t, s) -> IS_NOT_NULL.test(t) && comparator.compare(t, s);
     }
 
+    private static Integer sizeOf(TypedObject object) {
+        Object o = object.getValue();
+        if (o instanceof List) {
+            return List.class.cast(o).size();
+        }
+        if (o instanceof Map) {
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/querying/FilterOperations.java]: Support for String as well?

File: src/test/java/com/yahoo/bullet/querying/FilterOperationsTest.java [hunk 3]
```diff
@@ -337,6 +340,19 @@ public void testRegexLike() {
         Assert.assertFalse(FilterOperations.perform(RecordBox.get().add("id", "55001").getRecord(), clause));
     }
 
+    @Test
+    public void testSizeOf() {
```
Comment (akshaisarma) [inline src/test/java/com/yahoo/bullet/querying/FilterOperationsTest.java]: Move to after regex tests?

File: src/main/java/com/yahoo/bullet/parsing/Clause.java (hunk 1 above)
Comment (hbxie) [reply src/main/java/com/yahoo/bullet/parsing/Clause.java]: Done.

File: src/main/java/com/yahoo/bullet/querying/FilterOperations.java (hunk 2 above)
Comment (hbxie) [reply src/main/java/com/yahoo/bullet/querying/FilterOperations.java]: Done.

File: src/test/java/com/yahoo/bullet/querying/FilterOperationsTest.java (hunk 3 above)
Comment (hbxie) [reply src/test/java/com/yahoo/bullet/querying/FilterOperationsTest.java]: Done.

File: src/test/java/com/yahoo/bullet/querying/FilterOperationsTest.java [hunk 4]
```diff
@@ -352,6 +355,21 @@ public void testBadRegex() {
...
+        FilterClause clause = getFieldFilter("id", SIZE_OF, "1", "2");
+        Assert.assertFalse(FilterOperations.perform(RecordBox.get().getRecord(), clause));
+        Assert.assertTrue(FilterOperations.perform(RecordBox.get().add("id", 1).getRecord(), clause));
+        Assert.assertTrue(FilterOperations.perform(RecordBox.get().add("id", "12").getRecord(), clause));
+        Assert.assertFalse(FilterOperations.perform(RecordBox.get().add("id", "123").getRecord(), clause));
+        Map<String, Object> map = new HashMap<>();
```
Comment (akshaisarma) [inline src/test/java/com/yahoo/bullet/querying/FilterOperationsTest.java]: Use Collections.singletonMap instead?

File: src/main/java/com/yahoo/bullet/querying/FilterOperations.java [hunk 5]
```diff
@@ -71,6 +73,7 @@
         COMPARATORS.put(Clause.Operation.LESS_EQUALS, isNotNullAnd(LE));
     }
     static final Comparator<Pattern> REGEX_LIKE = isNotNullAnd(RLIKE);
+    static final Comparator<TypedObject> SIZE_OF = isNotNullAnd(SIZEOF);
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/querying/FilterOperations.java]: Do we want to think of SIZEOF(null) == 0 as true or false? Right now, it's false.

File: src/test/java/com/yahoo/bullet/querying/FilterOperationsTest.java [hunk 6]
```diff
@@ -352,6 +355,21 @@ public void testBadRegex() {
...
+        map.put("1", 1);
+        Assert.assertTrue(FilterOperations.perform(RecordBox.get().addList("id", map).getRecord(), clause));
```
Comment (akshaisarma) [inline src/test/java/com/yahoo/bullet/querying/FilterOperationsTest.java]: Can we have a test for an empty map and list as well?

File: src/test/java/com/yahoo/bullet/querying/FilterOperationsTest.java (hunk 4 above)
Comment (hbxie) [reply src/test/java/com/yahoo/bullet/querying/FilterOperationsTest.java]: Done.

File: src/test/java/com/yahoo/bullet/querying/FilterOperationsTest.java (hunk 6 above)
Comment (hbxie) [reply src/test/java/com/yahoo/bullet/querying/FilterOperationsTest.java]: Added.

File: src/main/java/com/yahoo/bullet/querying/FilterOperations.java (hunk 5 above)
Comment (hbxie) [reply src/main/java/com/yahoo/bullet/querying/FilterOperations.java]: We can. But I saw RLIKE(null, ".*") return false, so I added isNotNullAnd.

Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/querying/FilterOperations.java]: Ok, let's leave it as false unless @NathanSpeidel has a different opinion.

**Pull request merged**